bin/bench

```diff
@@ -0,0 +1,278 @@
+    at_exit { empty_temp_dir }
+  end
+
+  def refresh_temp_dir
+    empty_temp_dir
+    fill_temp_dir
+  end
+
+  module RevisionMethods
+    module_function
+
+    def current_branch
+      @current_branch ||= `cat .git/HEAD | cut -d/ -f3,4,5`.chomp
+    end
+
+    def revisions(start_ref, end_ref)
+      cmd = "git rev-list --reverse #{start_ref}..#{end_ref}"
+      `#{cmd}`.chomp.split("\n")
+    end
+
+    def checkout_ref(ref)
+      debug { `git checkout #{ref}`.chomp }
+      abort "Checkout failed: #{ref}, #{$CHILD_STATUS.exitstatus}" unless $CHILD_STATUS.success?
+    end
+
+    def revision_description(rev)
+      `git log --oneline -1 #{rev}`
+    end
+
+    def bundle
+      `rm -f Gemfile.lock; bundle check || bundle --local --quiet || bundle --quiet`
+    end
+
+    def clean_head
+      system('git reset --hard --quiet')
+    end
+  end
+  include RevisionMethods
+
+  def benchmark_refs(ref1: nil, ref2: nil)
+    ref0 = current_branch
+    ref1 ||= current_branch
+    ref2 ||= 'master'
+
+    reports = {}
+    reports[ref1] = run_benchmark_at_ref(ref1)
+    reports[ref2] = run_benchmark_at_ref(ref2)
+
+    checkout_ref(ref0)
+
+    reports
+  rescue Exception # rubocop:disable Lint/RescueException
+    checkout_ref(ref0)
+    raise
+  ensure
+    return reports
+  end
+
+  def benchmark_revisions(ref1: nil, ref2: nil)
+    ref0 = current_branch
+    ref1 ||= current_branch
+    ref2 ||= 'master'
+    reports = {}
+
+    revisions(ref1, ref2).each do |rev|
+      STDERR.puts "Checking out: #{revision_description(rev)}"
+
+      reports[rev] = run_benchmark_at_ref(rev)
+      clean_head
+    end
+    checkout_ref(ref0)
+    debug { "OK for all revisions!" }
+    reports
+  rescue Exception # rubocop:disable Lint/RescueException
+    STDERR.puts $!.message
+    checkout_ref(ref0)
+    raise
+  ensure
+    return reports
+  end
+
+  def run_benchmark_at_ref(ref)
+    checkout_ref(ref)
+    run_benchmark
+  end
+
+  def run_benchmark
+    bundle
+    restart_server
+    warmup
+    parse_measurement Benchmark.measure {
+      benchmark_tests
+    }
+  end
+
+  # TODO: benchmark these separately, using benchmark/ips.
+  # TODO: benchmark each with caching on and off
+  def benchmark_tests
+    100.times do
+      get_caching
+      get_non_caching
+    end
+  end
+
+  def restart_server
+    server_script = File.join(TMP_DIR_BASE, 'serve_dummy')
+    system("#{server_script} stop")
+    at_exit { system("#{server_script} stop") }
+    config_ru = Shellwords.shellescape(File.join(TMP_DIR_BASE, 'config.ru'))
+    pid = `CONFIG_RU=#{config_ru} #{server_script} start`.chomp
+    abort "No pid" if pid.empty?
+    pid = Integer(pid)
+    Process.kill(0, pid) # confirm running, else it raises
+  end
+
+  def warmup
+    assert_responses(get_caching, get_non_caching)
+    10.times do
+      get_caching
+      get_non_caching
+    end
+  end
+
+  # TODO: support /caching/{on,off}
+  def get_caching
+    get(url_base + "/caching")
+  end
```

Comment:
basically, the benchmark is a some combination or requests made against the server spawned at that ref.  This way, everything can be controlled from the benchmarking file.  e.g., if I wanted to run benchmark-ips, I would run that in here, not in the dummy controller. I think this makes it more flexible.

Also, cache toggling is now implemented in the server

```diff
+Rails.application.routes.draw do
 +  get '/status(/:on)' => 'post#render_cache_status'
 +  get '/caching(/:on)' => 'post#render_with_caching_serializer'
 +  get '/non_caching(/:on)' => 'post#render_with_non_caching_serializer'
 +end
```